File: src/models/deck.ts
import { Card, Suit, Rank, createCard, getAllCardTypes } from './card';

const COPIES_PER_CARD = 4;

export interface Deck {
  readonly cards: Card[];
  readonly drawPile: Card[];
  readonly discardPile: Card[];
}

export function createDeck(): Deck {
  const cards: Card[] = [];
  const types = getAllCardTypes();

  for (const cardType of types) {
    for (let i = 0; i < COPIES_PER_CARD; i++) {
      cards.push({
        ...cardType,
        id: `${cardType.rank}_${cardType.suit}_${i}`,
      });
    }
  }

  return {
    cards,
    drawPile: [...cards],
    discardPile: [],
  };
}

// Fisher-Yates shuffle
export function shuffleDeck(deck: Deck, seed?: number): Deck {
  const drawPile = [...deck.drawPile];

  let rng = seed !== undefined ? createSeededRandom(seed) : Math.random;

  for (let i = drawPile.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    [drawPile[i], drawPile[j]] = [drawPile[j], drawPile[i]];
  }

  return { ...deck, drawPile };
}

export function drawCard(deck: Deck): { card: Card; deck: Deck } {
  if (deck.drawPile.length === 0) {
    throw new Error('Nọc is empty - no cards left to draw');
  }

  const [card, ...remaining] = deck.drawPile;
  return {
    card,
    deck: { ...deck, drawPile: remaining },
  };
}

export function discardCard(deck: Deck, card: Card): Deck {
  return {
    ...deck,
    discardPile: [...deck.discardPile, card],
  };
}

export function topDiscard(deck: Deck): Card | null {
  if (deck.discardPile.length === 0) return null;
  return deck.discardPile[deck.discardPile.length - 1];
}

export function drawPileCount(deck: Deck): number {
  return deck.drawPile.length;
}

// Simple seeded PRNG (mulberry32)
function createSeededRandom(seed: number): () => number {
  let state = seed;
  return () => {
    state |= 0;
    state = (state + 0x6d2b79f5) | 0;
    let t = Math.imul(state ^ (state >>> 15), 1 | state);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

File: src/ChanDanGian.Core/Models/Deck.cs
namespace ChanDanGian.Core.Models;

public record Deck(List<Card> DrawPile, List<Card> DiscardPile)
{
    public int DrawPileCount => DrawPile.Count;
    public Card? TopDiscard => DiscardPile.Count > 0 ? DiscardPile[^1] : null;
}

public static class DeckFactory
{
    public const int CopiesPerCard = 4;

    public static Deck CreateDeck()
    {
        var cards = new List<Card>();
        int id = 0;

        foreach (Rank rank in Enum.GetValues<Rank>())
        {
            foreach (Suit suit in Enum.GetValues<Suit>())
            {
                for (int copy = 0; copy < CopiesPerCard; copy++)
                {
                    cards.Add(new Card($"c{id++}", rank, suit));
                }
            }
        }

        return new Deck(cards, new List<Card>());
    }

    public static Deck ShuffleDeck(Deck deck, int? seed = null)
    {
        var rng = seed.HasValue ? new Random(seed.Value) : new Random();
        var shuffled = deck.DrawPile.ToList();

        // Fisher-Yates shuffle
        for (int i = shuffled.Count - 1; i > 0; i--)
        {
            int j = rng.Next(i + 1);
            (shuffled[i], shuffled[j]) = (shuffled[j], shuffled[i]);
        }

        return deck with { DrawPile = shuffled };
    }

    public static (Card Card, Deck Deck) DrawCard(Deck deck)
    {
        if (deck.DrawPile.Count == 0)
            throw new InvalidOperationException("Nọc is empty");

        var card = deck.DrawPile[^1];
        var newPile = deck.DrawPile.Take(deck.DrawPile.Count - 1).ToList();
        return (card, deck with { DrawPile = newPile });
    }

    public static Deck DiscardCard(Deck deck, Card card)
    {
        var newDiscard = new List<Card>(deck.DiscardPile) { card };
        return deck with { DiscardPile = newDiscard };
    }
}
